### DISCUSSION TITLE
Generate a drift database based on an existing drift database

### DISCUSSION THREAD

**Discussion by miikaeru** [simolus3/drift#2352] [Q&A]
Suppose I generate the following simple Drift database with a todo table in a Flutter package, populate it with data, and export it to a `todo.db` file.

```dart
@DriftDatabase(
  tables: [
    Todos,
  ],
  include: {'tables.drift'},
)
class TodoDatabase extends _$TodoDatabase {
  TodoDatabase() : super(_openConnection());

  Future<List<Todo>> get allTodos => select(todos).get();

  @override
  int get schemaVersion => 1;
}

LazyDatabase _openConnection() {
  return LazyDatabase(() async {
    Directory current = Directory.current;
    final file = File(p.join(current.path, 'assets/todo.db'));
    return NativeDatabase.createInBackground(file);
  });
}
```

If I import this package and the existing `todo.db` file into a Flutter app, how do I generate a new Drift database that extends from the database containing the todo table?

E.g.
```dart
import 'package:todo/todo.dart';

@DriftDatabase(
  tables: [
    UserNotes
  ]
)
class LocalDatabase extends _$LocalDatabase with TodoDatabase {
  @override
  LocalDatabase() : super(_openConnection());

  @override
  int get schemaVersion => 1;
}

_LazyDatabase openConnection() {
  return LazyDatabase(() async {
    final dbFolder = await getApplicationDocumentsDirectory();
    final file = File(p.join(dbFolder.path, 'local.db'));

    if (!await file.exists()) {
      final blob = await rootBundle.load('assets/todo.db');
      final buffer = blob.buffer;
      await file.writeAsBytes(
          buffer.asUint8List(blob.offsetInBytes, blob.lengthInBytes));
    }

    return NativeDatabase.createInBackground(file);
  });
}
```

**Comment by simolus3** [comment]
In its default generation mode, drift generates code from the perspective of a single database. This makes this approach unfeasible unfortunately, as the row and companion classes for `Todo` would have to be generated in a part file for `LocalDatabase`.

In recent versions, drift added support for [modular code generation](https://drift.simonbinder.eu/docs/advanced-features/builder_options/#modular-code-generation), with which drift will generate code from the perspective of each table. With modular code generation (which would have to be enabled for both `package:todo` and the client package), code for the `Todo` table is generated inside the `todos` package. I probably wouldn't expose `TodoDatabase` then (it's also no longer necessary for code to be generated), but you could write a database accessor for the todo table. The main package could then add this accessor.

A (single package) example is available [here](https://github.com/simolus3/drift/tree/develop/examples/modular). I don't see a reason this shouldn't work for multiple packages as well, but I've never tried that setup. If you run into any issues with that, I'm happy to help resolve them. But I don't think this will realistically become possible in the default (non-modular) generation mode.
